### ISSUE TITLE
apply and refine do not respect opacity (in most but not all cases)

### ISSUE THREAD

**Issue by Armael** [coq/coq#9135]
`apply` seems to ignore opacity when no evars are involved
```coq
Definition t (n : nat) := n.
Global Opaque t.

Parameter P : nat -> Prop.

Goal forall n, P (t n) -> P n.
  intros n H. apply H. (* refine H also works *)
Qed.
```

However here, opacity is taken into account by `apply`, but still not by `refine`.
```coq
Definition t (n : nat) := n.
Global Opaque t.

Parameter P : nat -> Prop.

Goal (forall n, P (t n)) -> forall n, P n.
  intros H n. Fail apply H. refine (H _).
Qed.
```

**Comment by ppedrot** [comment]
Funny. The implementation of `refine` does not typecheck the term it is given against the conclusion, because the former comes fresh out of the pretyper with the type constraint and trusts it. But this is wrong, the pretyper doesn't seem to care about the transparent state!

**Comment by ppedrot** [comment]
And... this is not a bug, it's a feature! The behaviour was introduced by ab5d23b0a82fba784080946f6bb00dd123f64080, as preventing the unfolding of opaque terms in tactics would break CoRN. My natural tendency would be to think that this should be fixed in CoRN rather than hacked around in Coq, but I'll let @mattam82 comment.

**Comment by maximedenes** [comment]
> And... this is not a bug, it's a feature!

Well, this is clearly a bug an not a feature because it contradicts the documented behavior. Maybe it is a documentation bug, but then it should be fixed.

> The behaviour was introduced by [ab5d23b](https://github.com/coq/coq/commit/ab5d23b0a82fba784080946f6bb00dd123f64080)

Isn't this patch rather reverting an attempted fix?

**Comment by mattam82** [comment]
Indeed, it was a try to respect opacity that was defeated by the contribs
at the time, AFAIR. With our current CI it would be easier to fix the
problem and adapt the user developments we track accordingly.
Le mer. 26 déc. 2018 à 12:49, Maxime Dénès <notifications@github.com> a
écrit :

**Comment by ppedrot** [comment]
@maximedenes Obviously it's just a matter of phrasing. I believe that the correct thing to do is to make pretyping respect opacity once again, and fix the contribs. Easier said than done, but maybe @mattam82 is right that the CI is going to help. Who's opening the PR?

**Comment by Blaisorblade** [comment]
👍: I've also run into this and wondered for a while what I was doing wrong.

**Comment by Blaisorblade** [comment]
TIL: convertibility is documented to not fully respect `Opaque`:

> Opaque has also an effect on the conversion algorithm of Coq, telling it to delay the unfolding of a constant as much as possible when Coq has to check the conversion (see Section Conversion rules) of two distinct applied constants.

In "Ideal Coq", convertibility would instead *never* unfold opaque constants I suppose, just like (hopefully) it never unfolds proofs hidden by `Qed`?

**Comment by Blaisorblade** [comment]
`apply:` shares this problem. Minimized example from https://gitter.im/coq/coq?at=5e5e193ab873303e27779494.

```
From Coq Require Import ssreflect.
Definition opaque {A} (a : A) := a.
Opaque opaque.

Goal exists x, 0 = opaque x. 
  eexists.
  (* old unification respects opaqueness *)
  Fail apply eq_refl.
  (* new unification does not *)
  apply: eq_refl.
Qed.
```
This came up again in https://gitlab.mpi-sws.org/iris/iris/issues/301#note_46408, but it's not critical there.
